h2/src/main/org/h2/engine/Session.java

```diff
@@ -1832,41 +1862,82 @@ private long getStatementSavepoint() {
      * Start a new statement within a transaction.
      * @param command about to be started
      */
+    @SuppressWarnings("incomplete-switch")
     public void startStatementWithinTransaction(Command command) {
         Transaction transaction = getTransaction();
-        if(transaction != null) {
-            Set<MVMap<?, ?>> maps = null;
+        if (transaction != null) {
+            HashSet<MVMap<?, ?>> currentMaps = null, allMaps = null;
             if (command != null) {
-                maps = new HashSet<>();
                 Set<DbObject> dependencies = command.getDependencies();
+                currentMaps = new HashSet<>();
                 for (DbObject dependency : dependencies) {
                     if (dependency instanceof MVTable) {
-                        MVTable table = (MVTable) dependency;
-                        for (Index index : table.getIndexes()) {
-                            if (index instanceof MVIndex) {
-                                MVIndex index1 = (MVIndex) index;
-                                MVMap<?, ?> map = index1.getMVMap();
-                                maps.add(map);
+                        addTableToDependencies((MVTable) dependency, currentMaps);
+                    }
+                }
+                switch (isolationLevel) {
+                case REPEATABLE_READ: {
+                    allMaps = new HashSet<>();
+                    HashSet<MVTable> processed = new HashSet<>();
+                    for (DbObject dependency : dependencies) {
+                        if (dependency instanceof MVTable) {
+                            addTableToDependencies((MVTable) dependency, dependencies, allMaps, processed);
```

Comment:
Why do we care about tables referenced by integrity constrains in this case only?

Reply:
Lowest two levels use snapshots of tables only around the current command, they don't need any additions. Serializable performs full snapshot, there is nothing to add.

Repeatable read uses partial snapshots and appends new tables to it for each new command if it needs something new. To avoid visible violations of FK constraints they are appended to the snapshot. I can remove this code if nobody cares.